src/models/state/wallet/wallet_state.rs

```diff
@@ -3734,4 +3930,383 @@ mod tests {
+                    let public_announcement = key
+                        .to_address()
+                        .unwrap()
+                        .generate_public_announcement(utxo_notification_payload);
+                    public_announcements.push(public_announcement);
+
+                    let addition_record = commit(
+                        Tip5::hash(&utxo),
+                        sender_randomness,
+                        receiver_preimage.hash(),
+                    );
+                    if select {
+                        addition_records.push(addition_record);
+                    }
+
+                    (
+                        select,
+                        kt,
+                        ri,
+                        ai,
+                        utxo,
+                        sender_randomness,
+                        receiver_preimage,
+                    )
+                })
+                .collect_vec();
+
+            let new_kernel = TransactionKernelModifier::default()
+                .public_announcements(public_announcements)
+                .outputs(addition_records)
+                .modify(kernel);
+
+            // scan
+            let caught_utxos = wallet_state
+                .scan_for_utxos_announced_to_future_keys(num_future_keys, &new_kernel)
+                .await
+                .collect_vec();
+
+            // filter master list according to expectation
+            let filtered_utxos = all_utxos
+                .into_iter()
+                .filter(|(select, _, _, _, _, _, _)| {
+                    if !*select {
+                        println!("rejecting UTXO because not selected");
+                    }
+                    *select
+                })
+                .filter(|(_, _, relative_index, _, _, _, _)| {
+                    let index_in_range = *relative_index < num_future_keys;
+                    if !index_in_range {
+                        println!(
+                            "rejecting UTXO because index {} >= {}",
+                            *relative_index, num_future_keys
+                        );
+                    }
+                    index_in_range
+                })
+                .map(
+                    |(
+                        _,
+                        keytype,
+                        _,
+                        absolute_index,
+                        utxo,
+                        sender_randomness,
+                        receiver_preimage,
+                    )| {
+                        (
+                            keytype,
+                            absolute_index,
+                            IncomingUtxo {
+                                utxo,
+                                sender_randomness,
+                                receiver_preimage,
+                            },
+                        )
+                    },
+                )
+                .collect_vec();
```

Comment:
I don't think you're doing yourself any favors by using iterator combinators here. The same is true higher up.

```suggestion
            let mut filtered_utxos = Vec::new();
            for (
                select,
                keytype,
                relative_index,
                absolute_index,
                utxo,
                sender_randomness,
                receiver_preimage,
            ) in all_utxos
            {
                if !select {
                    println!("rejecting UTXO because not selected");
                    continue;
                }
                if relative_index >= num_future_keys {
                    println!("rejecting UTXO because index {relative_index} >= {num_future_keys}");
                    continue;
                }
                let incoming_utxo = IncomingUtxo {
                    utxo,
                    sender_randomness,
                    receiver_preimage,
                };
                filtered_utxos.push((keytype, absolute_index, incoming_utxo));
            }
```

If you want to make the result immutable again, you can append `let filtered_utxos = filtered_utxos;`.

Reply:
It might even be beneficial to introduce a helper struct within this test.

```rust
#[traced_test]
#[tokio::test]
async fn scan_for_utxos_announced_to_future_keys_behaves() {
    struct MyTestHelper {
        is_selected: bool,
        utxo: Utxo,
        // …
    }

    // …
}
```